pkg/controllers/node/node_controller.go

```diff
@@ -182,6 +198,105 @@ func (c *NodeController) syncDelete() error {
 	return c.syncDeleteEtcd()
 }
 
+// isAutoHostendpoint determines if the given hostendpoint is managed by
+// kube-controllers.
+func isAutoHostendpoint(h *api.HostEndpoint) bool {
+	v, ok := h.Labels[hepCreatedLabelKey]
+	return ok && v == hepCreatedLabelValue
+}
+
+// createHostendpoint creates an auto hostendpoint for the specified node.
+func (c *NodeController) createHostendpoint(n *api.Node) (*api.HostEndpoint, error) {
+	hep := c.generateHostendpointFromNode(n)
+
+	time.Sleep(c.rl.When(RateLimitCalicoCreate))
+	res, err := c.calicoClient.HostEndpoints().Create(c.ctx, hep, options.SetOptions{})
+	if err != nil {
+		log.Warnf("could not create hostendpoint for node: %v", err)
+		return nil, err
+	}
+	c.rl.Forget(RateLimitCalicoCreate)
+	return res, nil
+}
+
+// syncAllHostendpoints syncs all Calico nodes with their hostendpoints.
+// If hostendpoints don't already exist, they are created. Dangling auto
+// hostendpoints are deleted.
+func (c *NodeController) syncAllHostendpoints() {
+	log.Debug("syncing all hostendpoints")
+
+	// On failure, we retry a certain number of times.
+	for n := 0; n < 5; n++ {
+		nodesList, err := c.calicoClient.Nodes().List(c.ctx, options.ListOptions{})
+		if err != nil {
+			log.WithError(err).Info("could not list nodes, retrying")
+			time.Sleep(retrySleepTime)
+			continue
+		}
+
+		hepsList, err := c.calicoClient.HostEndpoints().List(c.ctx, options.ListOptions{})
+		if err != nil {
+			log.WithError(err).Info("could not list hostendpoints, retrying")
+			time.Sleep(retrySleepTime)
+			continue
+		}
+
+		// Filter the list of hostendpoints so that we are working only with auto
+		// hostendpoints created by kube-controllers.
+		heps := make(map[string]api.HostEndpoint)
+		for _, h := range hepsList.Items {
+			if isAutoHostendpoint(&h) {
+				heps[h.Spec.Node] = h
+			}
+		}
+
+		// Go through current nodes, creating hostendpoints for them if they don't exist.
+		for _, n := range nodesList.Items {
```

Comment:
Actually, I _think_ this will be covered by the Syncer - it will give us a full set of Nodes that exist in the data store when it starts up. So, I think the only thing the sync needs to do is just cleanup any HEPs that no longer belong... right?

Reply:
Yeah you're right